### PR TITLE
[Bugfix] Fix the LoRA weight sharding in ColumnParallelLinearWithLoRA

### PR DESCRIPTION
**Pull request by jeejeelee**
- [x] Add  comments for better understanding
- [x] Add some LoRA tests to distributed tests(refer to :https://github.com/vllm-project/vllm/pull/10352#issuecomment-2478081472)

### CODE REVIEW

File: tests/lora/test_chatglm3_tp.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        "SELECT name ,  country ,  age FROM singer ORDER BY age",
+    ]
+
+    output1 = do_sample(llm, chatglm3_lora_files, lora_id=1)
+    for i in range(len(expected_lora_output)):
+        assert output1[i] == expected_lora_output[i]
+    output2 = do_sample(llm, chatglm3_lora_files, lora_id=2)
+    for i in range(len(expected_lora_output)):
+        assert output2[i] == expected_lora_output[i]
+    cleanup_dist_env_and_memory()
```
Comment (jeejeelee) [inline tests/lora/test_chatglm3_tp.py:121]: @DarkLight1337   want to test TP with this model, but I can't even get it to pass locally. Could you help me check what might be wrong with my implementation, thanks

File: vllm/lora/layers.py [hunk 2]
```diff
@@ -508,14 +514,30 @@ def slice_lora_a(self, lora_a: torch.Tensor) -> torch.Tensor:
...
+        else:
+            tensor_model_parallel_rank = get_tensor_model_parallel_rank()
+            shard_size = self.output_dim
+            start_idx = tensor_model_parallel_rank * shard_size
+            end_idx = (tensor_model_parallel_rank + 1) * shard_size
+            lora_b = lora_b[:, start_idx:end_idx]
         return lora_b
 
     def slice_bias(self, bias: torch.Tensor) -> torch.Tensor:
+        # TODO: Fix the slicing logic of bias.
```
Comment (DarkLight1337) [inline vllm/lora/layers.py:540]: Maybe it's because you haven't implemented this yet?

File: vllm/lora/layers.py [hunk 3]
```diff
@@ -508,14 +514,30 @@ def slice_lora_a(self, lora_a: torch.Tensor) -> torch.Tensor:
...
+        if self.is_merged_col_linear:
+            tp_rank = get_tensor_model_parallel_rank()
+            shard_size = self.output_size // 2
+            offset = lora_b.shape[-1] // 2
+
+            left_weight = lora_b[:, tp_rank * shard_size:(tp_rank + 1) *
+                                 shard_size]
+            right_weigt = lora_b[:, offset + tp_rank * shard_size:offset +
+                                 (tp_rank + 1) * shard_size]
+            lora_b = torch.cat([left_weight, right_weigt], dim=1)
```
Comment (DarkLight1337) [inline vllm/lora/layers.py:528]:
```suggestion
            right_weight = lora_b[:, offset + tp_rank * shard_size:offset +
                                 (tp_rank + 1) * shard_size]
            lora_b = torch.cat([left_weight, right_weight], dim=1)
```

File: vllm/lora/layers.py (hunk 2 above)
Comment (jeejeelee) [reply vllm/lora/layers.py:540]: No, I don't plan to address the bias slicing logic in this PR. I have doubts about the bias implementation, as I haven't fully understood it yet.

File: vllm/lora/fully_sharded_layers.py [hunk 4]
```diff
@@ -45,10 +45,24 @@ class ColumnParallelLinearWithShardedLoRA(ColumnParallelLinearWithLoRA):
...
+        # MergedColumnParallelLinear.
+        if self.is_merged_col_linear:
+            tp_rank = get_tensor_model_parallel_rank()
+            shard_size = self.lora_a_stacked.shape[2] // 2
+            offset = lora_a.shape[-1] // 2
+            left_weight = lora_a[:, tp_rank * shard_size:(tp_rank + 1) *
+                                 shard_size]
+            right_weigt = lora_a[:, offset + tp_rank * shard_size:offset +
+                                 (tp_rank + 1) * shard_size]
+            lora_a = torch.cat([left_weight, right_weigt], dim=1)
```
Comment (DarkLight1337) [inline vllm/lora/fully_sharded_layers.py:58]:
```suggestion
            right_weight = lora_a[:, offset + tp_rank * shard_size:offset +
                                 (tp_rank + 1) * shard_size]
            lora_a = torch.cat([left_weight, right_weight], dim=1)
```

Comment (DarkLight1337) [reply vllm/lora/fully_sharded_layers.py:58]: Also here

File: vllm/lora/layers.py (hunk 2 above)
Comment (jeejeelee) [reply vllm/lora/layers.py:540]: For this test [script](https://github.com/vllm-project/vllm/blob/main/tests/lora/test_llama.py), even with 4 GPUs, the test still fails. The main issue is that the test keeps hanging."

File: tests/lora/test_chatglm3_tp.py (hunk 1 above)
Comment (DarkLight1337) [reply tests/lora/test_chatglm3_tp.py:121]: I'm not familiar about the layer implementation of `MergedColumnParallelLinear`. Maybe you can print out the slice indices and see if they make sense.

File: vllm/lora/layers.py (hunk 2 above)
Comment (DarkLight1337) [reply vllm/lora/layers.py:540]: If it's about the test hanging, maybe some of the workers failed to call `torch.distributed.all_gather`.

File: tests/lora/test_chatglm3_tp.py (hunk 1 above)
Comment (jeejeelee) [reply tests/lora/test_chatglm3_tp.py:121]: For this test [script](https://github.com/vllm-project/vllm/blob/main/tests/lora/test_llama.py), even with 4 GPUs, the test still fails. The main issue is that the test keeps hanging. Without considering unit tests, I've verified that TP=1/2/4 works correctly.

Comment (DarkLight1337) [reply tests/lora/test_chatglm3_tp.py:121]: Does it only fail for the TP4 test?

Comment (jeejeelee) [reply tests/lora/test_chatglm3_tp.py:121]: Just run
```shell 
pytest  test_chatglm3_tp.py
```

Comment (jeejeelee) [reply tests/lora/test_chatglm3_tp.py:121]: Okay, I will try

**Comment by jeejeelee** [comment]
@DarkLight1337  I can now complete multi-GPU unit tests locally. Could you tell me where I should add multi-GPU LoRA tests? Is it correct  [here](https://github.com/vllm-project/vllm/blob/main/.buildkite/test-pipeline.yaml)?

**Comment by jeejeelee** [comment]
> Yeah, you should update that file.

Thanks, could you plz look this PR again, thanks

**Comment by DarkLight1337** [comment]
Can you make these tests run with the existing 4 GPU tests?

**Comment by jeejeelee** [comment]
> Can you make these tests run with the existing 4 GPU tests?

Done, plz look at this again, thanks for your hard work

File: .buildkite/test-pipeline.yaml [hunk 5]
```diff
@@ -475,18 +475,23 @@ steps:
   - pytest -v -s distributed/test_pp_cudagraph.py
   - pytest -v -s distributed/test_pipeline_parallel.py
 
-- label: LoRA Long Context (Distributed) # 11min
-  # This test runs llama 13B, so it is required to run on 4 GPUs.
+- label: LoRA TP Test  (Distributed)
```
Comment (DarkLight1337) [inline .buildkite/test-pipeline.yaml:478]:
```suggestion
- label: LoRA TP Test (Distributed)
```

**Review by DarkLight1337 — APPROVED** [review]
Otherwise looks good, thanks for the fix!

File: vllm/model_executor/models/chatglm.py [hunk 6]
```diff
@@ -747,7 +747,7 @@ def __new__(
         config = vllm_config.model_config.hf_config
         # Initialize VL
         if hasattr(config, "visual"):
-            return ChatGLM(vllm_config=vllm_config, prefix=prefix)
+            return ChatGLMV(vllm_config=vllm_config, prefix=prefix)
         # Initialize LLM
         else:
-            return ChatGLMV(vllm_config=vllm_config, prefix=prefix)
+            return ChatGLM(vllm_config=vllm_config, prefix=prefix)
```
Comment (jeejeelee) [inline vllm/model_executor/models/chatglm.py:753]: By the way, fix the bug introduced by adding LoRA support for GLM.

**Comment by DarkLight1337** [comment]
https://github.com/vllm-project/vllm/pull/10581 should fix the CI failure, please merge from main again.

**Comment by DarkLight1337** [comment]
Can you check whether the examples test failure is related to this PR?

**Comment by jeejeelee** [comment]
I'm outside now, I'll check later​​​​​​​​​​​​​​​​

**Comment by jeejeelee** [comment]
@DarkLight1337  DCO always 

> Can you check whether the examples test failure is related to this PR?

It seems the example tests triggered in CI are not related to this PR. Could you tell me which exampletest failed? I can try to reproduce it locally.

**Comment by DarkLight1337** [comment]
See here for detailed logs: https://buildkite.com/vllm/ci-aws/builds/11689#01935755-0382-4a69-8f86-413cdb8a12c0

**Comment by jeejeelee** [comment]
> See here for detailed logs: https://buildkite.com/vllm/ci-aws/builds/11689#01935755-0382-4a69-8f86-413cdb8a12c0

Ah, I actually already looked at it, but I couldn't figure out which example failed, that's why I'm asking you.

**Comment by jeejeelee** [comment]
> See here for detailed logs: https://buildkite.com/vllm/ci-aws/builds/11689#01935755-0382-4a69-8f86-413cdb8a12c0

It seems to be related to https://github.com/vllm-project/vllm/blob/main/.buildkite/test-pipeline.yaml#L192, is that right?

**Comment by DarkLight1337** [comment]
Seems like it.

**Comment by jeejeelee** [comment]
> Seems like it.

I can reproduce this failure in this PR branch ,I am investigating the root cause now.

**Comment by jeejeelee** [comment]
> > Seems like it.
> 
> I can reproduce this failure in this PR branch ,I am investigating the root cause now.

I also can reproduce this failure in the  latest main branch

**Pull request merged**